lib/puppet/application/resource.rb

```diff
@@ -139,6 +148,20 @@ def main
 
       raise _("Editing with Yaml output is not supported") if options[:edit] and options[:to_yaml]
 
+      if options[:target]
+        require 'puppet/util/network_device/config'
+        target_device = Puppet::Util::NetworkDevice::Config.device(options[:target])
+        if target_device.nil?
+          raise _("Invalid target '%{target}'") % { target: options[:target]}
+        end
+        # Facter(:url) is used by the device module loaded via
+        # Puppet::Resource.indirection.search in find_or_save_resources()
+        url = target_device.url.to_s
+        Facter.add(:url) do
```

Comment:
The fact name seems overly generic. If it's only for remote devices, suggest something like `:device_url`, etc

Reply:
Yes, the fact name is too generic, but was chosen years ago :( ...

https://puppet.com/blog/managing-f5-big-ip-network-devices-puppet (search for FACTER_url)

... and is implemented in multiple device modules:

https://github.com/f5devcentral/f5-puppet/blob/master/lib/puppet/provider/f5.rb#L16
https://forge.puppet.com/puppetlabs/netscaler#tips-and-tricks

Also, https://tickets.puppetlabs.com/browse/MODULES-1751

Reply:
(I tried to identify an alternate mechanism to pass the device object, but was unable navigate indirection.)